File: examples/checks/docking_abort_token.py
# Independent Python checks for the docking_abort_token example.
from __future__ import annotations

import re

from .common import run_checks


def run(ctx):
    data = ctx.load_input()
    media = data["media"]
    names = [m["name"] for m in media]
    copy_tasks = [(a, b) for a in names for b in names if a != b]
    serial = data["expected"]["serial"].split("->")
    parallel_source = data["expected"]["parallelSource"]
    reported_copy = re.search(r"possible copy tasks : (\d+)", ctx.answer)
    reported_copy_count = int(reported_copy.group(1)) if reported_copy else None

    checks = [
        ("all four media carry the same AbortBit variable", len(media) == 4 and data["variable"] == "AbortBit"),
        ("each medium has distinct zero and one states", all(m["zero"] != m["one"] for m in media)),
        ("the directed copy-task count is recomputed as n*(n-1)", reported_copy_count == len(copy_tasks) == 12),
        ("the expected serial witness uses known media in order", all(item in names for item in serial) and f"{serial[0]} -> {serial[1]} -> {serial[2]}" in ctx.answer),
        ("the serial witness is backed by two legal copy/measure edges", (serial[0], serial[1]) in copy_tasks and (serial[1], serial[2]) in copy_tasks),
        ("the expected parallel source can fan out to two other media", parallel_source in names and sum(1 for a, _b in copy_tasks if a == parallel_source) >= 2 and "parallel witness : flightPLC -> radioFrame and auditDisplay" in ctx.answer),
        ("the quantum seal is separate from the classical AbortBit variable", data["superinformationMedium"]["variable"] != data["variable"]),
        ("the answer blocks universal cloning and unrestricted audit fan-out for the seal", "cannot be universally cloned" in ctx.answer and "unrestricted audit fan-out" in ctx.answer),
    ]
    return run_checks(checks)

File: examples/checks/docking_abort_token.go
package main

import (
	"strings"
)

func checkDocking(ctx *Context) []Check {
	d := ctx.M()
	media := maps(d["media"])
	names := []string{}
	for _, m := range media {
		names = append(names, str(m["name"]))
	}
	copyCount := len(names) * (len(names) - 1)
	reported := grabInt(ctx.Answer, `possible copy tasks : (\d+)`)
	serial := strings.Split(str(asMap(d["expected"])["serial"]), "->")
	parallel := str(asMap(d["expected"])["parallelSource"])
	nameSet := setOf(names)
	serialKnown := true
	for _, x := range serial {
		serialKnown = serialKnown && nameSet[x]
	}
	distinct := true
	for _, m := range media {
		distinct = distinct && str(m["zero"]) != str(m["one"])
	}
	super := asMap(d["superinformationMedium"])
	return []Check{{"all four media carry the same AbortBit variable", len(media) == 4 && str(d["variable"]) == "AbortBit"}, {"each medium has distinct zero and one states", distinct}, {"the directed copy-task count is recomputed as n*(n-1)", reported == copyCount && copyCount == 12}, {"the expected serial witness uses known media in order", serialKnown && contains(ctx.Answer, serial[0]+" -> "+serial[1]+" -> "+serial[2])}, {"the serial witness is backed by two legal copy/measure edges", len(serial) >= 3 && serial[0] != serial[1] && serial[1] != serial[2]}, {"the expected parallel source can fan out to two other media", nameSet[parallel] && copyCount/len(names) >= 3 && contains(ctx.Answer, "parallel witness : flightPLC -> radioFrame and auditDisplay")}, {"the quantum seal is separate from the classical AbortBit variable", str(super["variable"]) != str(d["variable"])}, {"the answer blocks universal cloning and unrestricted audit fan-out for the seal", contains(ctx.Answer, "cannot be universally cloned") && contains(ctx.Answer, "unrestricted audit fan-out")}}
}
